File: circles-frontend/src/lib/grading.ts
// Mirrors circles-backend/app/services/grading.py's grade_question — kept in
// sync manually since the two apps share no code. Used for the results/review
// view and the local-fallback scorer, both of which need per-question-type
// correctness without waiting on the server's aggregate score.
import { Question, QuestionType, QuizAnswerValue, getQuestionType } from "./api";

function normalize(s: string): string {
  return s.trim().toLowerCase().split(/\s+/).join(" ");
}

export function isAnswerCorrect(question: Question, submitted: QuizAnswerValue | undefined): boolean {
  const qtype: QuestionType = getQuestionType(question);

  if (qtype === "multiple_choice" || qtype === "true_false") {
    return submitted === (question as { correct_answer: string }).correct_answer;
  }

  if (qtype === "fill_in_blank") {
    if (typeof submitted !== "string") return false;
    const q = question as { accepted_answers: string[]; correct_answer: string };
    const accepted = q.accepted_answers?.length ? q.accepted_answers : [q.correct_answer];
    const acceptedNorm = new Set(accepted.filter((a) => typeof a === "string").map(normalize));
    return acceptedNorm.has(normalize(submitted));
  }

  if (qtype === "matching") {
    if (!submitted || typeof submitted !== "object") return false;
    const q = question as { pairs: { left: string; right: string }[] };
    const pairs = q.pairs ?? [];
    if (pairs.length === 0 || Object.keys(submitted).length !== pairs.length) return false;
    return pairs.every((p) => submitted[p.left] === p.right);
  }

  return false;
}

File: circles-backend/app/services/grading.py
from typing import Any


def grade_question(question: dict, submitted: Any) -> bool:
    """True iff `submitted` is a fully-correct answer for `question`.

    Missing/unrecognized question_type defaults to "multiple_choice" so
    pre-existing rows (generated before question_type existed) keep grading
    the way they always have.
    """
    qtype = question.get("question_type") or "multiple_choice"
    if qtype in ("multiple_choice", "true_false"):
        return submitted == question.get("correct_answer")
    if qtype == "fill_in_blank":
        return _grade_fill_in_blank(question, submitted)
    if qtype == "matching":
        return _grade_matching(question, submitted)
    return False


def _normalize(s: str) -> str:
    return " ".join(s.strip().lower().split())


def _grade_fill_in_blank(question: dict, submitted: Any) -> bool:
    if not isinstance(submitted, str):
        return False
    accepted = question.get("accepted_answers") or (
        [question["correct_answer"]] if question.get("correct_answer") else []
    )
    accepted_norm = {_normalize(a) for a in accepted if isinstance(a, str)}
    return _normalize(submitted) in accepted_norm


def _grade_matching(question: dict, submitted: Any) -> bool:
    if not isinstance(submitted, dict):
        return False
    pairs = question.get("pairs") or []
    if not pairs or len(submitted) != len(pairs):
        return False
    return all(submitted.get(p.get("left")) == p.get("right") for p in pairs)
